metagraph/tests/test_vector_algorithm.cpp

```diff
@@ -123,7 +123,162 @@ TEST(IntVector, call_nonzeros_sparse) {
+                }
+            }
+        }
+
+        for (size_t i = 0; i < vector.size(); ++i) {
+            EXPECT_EQ(vector[i], atomic_fetch(vector_relaxed, i, mu, __ATOMIC_ACQUIRE));
+            EXPECT_EQ(vector[i], atomic_fetch(vector_seq_cst, i, mu));
+        }
+
+        for (size_t begin = 0; begin < vector.size(); ++begin) {
+            for (size_t end = begin; end < vector.size(); ++end) {
+                uint64_t it = begin;
+                call_nonzeros(vector, begin, end, [&](auto i, auto count) {
+                    EXPECT_LE(begin, i);
+                    EXPECT_GT(end, i);
+                    while (vector[it] == 0) {
+                        EXPECT_GT(i, it);
+                        ++it;
+                        ASSERT_GT(end, it);
+                    }
+
+                    EXPECT_EQ(vector[i], count);
+                    EXPECT_EQ(it++, i);
+                });
+                for (; it < end; ++it) {
+                    EXPECT_EQ(0u, vector[it]);
+                }
+            }
+        }
+    }
+}
+
+TEST(IntVector, call_nonzeros_sparse_atomic_add) {
+    for (size_t w = 1; w <= 64; ++w) {
+        sdsl::int_vector<> vector_relaxed = aligned_int_vector(200, 0, w, 16);
+        sdsl::int_vector<> vector_seq_cst = aligned_int_vector(200, 0, w, 16);
+        sdsl::int_vector<> vector(200, 0, w);
+        ASSERT_LT(65u, vector.size());
+
+        std::mutex mu;
+        std::mutex mu2;
+        #pragma omp parallel for num_threads(3) schedule(dynamic)
+        for (size_t j = 0; j < 3; ++j) {
+            for (size_t i = j; i < vector.size(); i += 3) {
+                if (i % 130 == 1) {
+                    EXPECT_EQ(0u, atomic_fetch_and_add(vector_relaxed, i, 1, mu, __ATOMIC_RELAXED));
+                    EXPECT_EQ(0u, atomic_fetch_and_add(vector_seq_cst, i, 1, mu));
+                    std::lock_guard<std::mutex> lock(mu2);
+                    ++vector[i];
+                }
+            }
+        }
+
+        for (size_t i = 0; i < vector.size(); ++i) {
+            EXPECT_EQ(vector[i], atomic_fetch(vector_relaxed, i, mu, __ATOMIC_ACQUIRE));
+            EXPECT_EQ(vector[i], atomic_fetch(vector_seq_cst, i, mu));
+        }
+
+        for (size_t begin = 0; begin < vector.size(); ++begin) {
+            for (size_t end = begin; end < vector.size(); ++end) {
+                uint64_t it = begin;
+                call_nonzeros(vector, begin, end, [&](auto i, auto count) {
+                    EXPECT_LE(begin, i);
+                    EXPECT_GT(end, i);
+                    while (vector[it] == 0) {
+                        EXPECT_GT(i, it);
+                        ++it;
+                        ASSERT_GT(end, it);
+                    }
+
+                    EXPECT_EQ(vector[i], count);
+                    EXPECT_EQ(it++, i);
+                });
+                for (; it < end; ++it) {
+                    EXPECT_EQ(0u, vector[it]);
+                }
+            }
+        }
+    }
+}
```

Comment:
any reason to throw all vector_relaxed, vector_seq_cst, vector into one test?

Reply:
I changed it so that now iterates through different memorders instead